tla/consistency/SingleNode.tla

```diff
@@ -0,0 +1,122 @@
+---- MODULE SingleNode ----
+\* A lightweight specification to define the externally visible behaviour of CCF
+\* This specification has been inspired by https://github.com/tlaplus/azure-cosmos-tla
+\* Where possible, naming should be consistent with https://microsoft.github.io/CCF/main/index.html
+\* SingleNode considers a single node CCF service, so no view changes
+
+EXTENDS ExternalHistory
+
+\* Upper bound of the number of client events
+\* Note that this abstract specification does not model CCF nodes so there's no
+\* constant for the number of nodes
+CONSTANT HistoryLimit
+
+\* Abstract ledger views that contains only client transactions (no signatures)
+\* Indexed by view, each ledger is the ledger associated with leader of that view 
+\* In practice, the ledger of every CCF node is one of these or a prefix for one of these
+\* This could be switched to a tree which can represent forks more elegantly
+VARIABLES ledgerViews
+
+LedgerTypeOK ==
+    \A view \in DOMAIN ledgerViews:
+        \A seqnum \in DOMAIN ledgerViews[view]:
+            \* Each ledger entry is tuple containing a view and tx
+            \* The ledger entry index is the sequence number
+            /\ ledgerViews[view][seqnum].view \in Views
+            /\ ledgerViews[view][seqnum].tx \in Txs
+
+\* In this abstract version of CCF's consensus layer, each ledger is append-only
+LedgersMonoProp ==
+    [][\A view \in DOMAIN ledgerViews: IsPrefix(ledgerViews[view], ledgerViews[view]')]_ledgerViews
+
+vars == << history, ledgerViews >>
+
+TypeOK ==
+    /\ HistoryTypeOK
+    /\ LedgerTypeOK
+
+Init ==
+    /\ history = <<>>
+    /\ ledgerViews = [ x \in {1} |-> <<>>]
+
+IndexOfLastRequested ==
+    SelectLastInSeq(history, LAMBDA e : e.type \in {RwTxRequest, RoTxRequest})
+
+NextRequestId ==
+    IF IndexOfLastRequested = 0 THEN 0 ELSE history[IndexOfLastRequested].tx+1
+
+\* Submit new read-write transaction
+\* This could be extended to add a notion of session and then check for session consistency
+RwTxRequestAction ==
+    /\ Len(history) < HistoryLimit
+    /\ history' = Append(
+        history, 
+        [type |-> RwTxRequest, tx |-> NextRequestId]
+        )
+    /\ UNCHANGED ledgerViews
+
+\* Execute a read-write transaction
+RwTxExecuteAction ==
+    /\ \E i \in DOMAIN history :
+        /\ history[i].type = RwTxRequest
+        \* Check transaction has not already been added a ledger
+        /\ \A view \in DOMAIN ledgerViews: 
+            {seqnum \in DOMAIN ledgerViews[view]: 
+                history[i].tx = ledgerViews[view][seqnum].tx} = {}
```

Comment:
Is this just an empty-set check? I think `Cardinality(...) = 0` is more readably an activation condition, but I'll defer to people with more TLA+ experience.

Reply:
Most probably consider `... = {}` idiomatic TLA+.